configuration/config2/EnvironmentNode.hpp

```diff
@@ -0,0 +1,81 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef _CONFIG2_ENVIRONMENTNODE_HPP_
+#define _CONFIG2_ENVIRONMENTNODE_HPP_
+
+#include <memory>
+#include <map>
+#include "SchemaItem.hpp"
+#include "EnvironmentValue.hpp"
+#include "SchemaValue.hpp"
+#include "Status.hpp"
+#include "ValueDef.hpp"
+#include "platform.h"
+
+
+class DECL_EXPORT EnvironmentNode : public std::enable_shared_from_this<EnvironmentNode>
+{
+    public:
+
+        EnvironmentNode(const std::shared_ptr<SchemaItem> &pCfgItem, const std::string &elemName, const std::shared_ptr<EnvironmentNode> &pParent = nullptr) :
+            m_pSchemaItem(pCfgItem), m_name(elemName), m_pParent(pParent) { }
+        ~EnvironmentNode() { }
+        const std::string &getName() const { return m_name;  }
+        void addChild(std::shared_ptr<EnvironmentNode> pNode);
+        bool removeChild(std::shared_ptr<EnvironmentNode> pNode);
+        std::vector<std::shared_ptr<EnvironmentNode>> getChildren(const std::string &name="") const;
+        std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> getChildrenByName() const;
+        std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> getChildrenByConfigType() const;
+        bool hasChildren() const { return m_children.size() != 0; }
+        int getNumChildren() const { return m_children.size(); }
+        std::shared_ptr<EnvironmentNode> getParent() const;
+        void addAttribute(const std::string &name, std::shared_ptr<EnvironmentValue> pValue);
+        void setAttributeValues(const std::vector<ValueDef> &values, Status &status, bool allowInvalid, bool forceCreate);
+        void setAttributeValue(const std::string &name, const std::string &value, Status &status, bool allowInvalid=false, bool forceCreate=false);   // candidate for a variant?
+        std::string getAttributeValue(const std::string &name) const;                                  // candidate for a variant?
+        void addMissingAttributesFromConfig();
+        bool setNodeValue(const std::string &value, Status &status, bool force = false);
+        void setNodeEnvValue(const std::shared_ptr<EnvironmentValue> &pEnvValue) { m_pNodeValue = pEnvValue;  }
+        const std::shared_ptr<EnvironmentValue> &getNodeEnvValue() const { return m_pNodeValue;  }
+        bool isNodeValueSet() const { return m_pNodeValue != nullptr; }
+        std::vector<std::shared_ptr<EnvironmentValue>> getAttributes() const;
+        const std::shared_ptr<EnvironmentValue> getAttribute(const std::string &name) const;
+        bool hasAttribute(const std::string &name) const { return m_attributes.find(name) != m_attributes.end();  }
+        bool hasAttributes() const { return m_attributes.size() != 0; }
+        void setId(const std::string &id) { m_id = id; }
```

Comment:
Not sure how these Ids are supposed to work.  They seem to be generated incremental numbers but you report them in error messages.  As soon as the xml changes they will change.

Reply:
In your test you also seem to use them for lookup, but again they will change, no?

Reply:
Test.cpp is hardcoded test code. The ID values can change with an environment is loaded, but are immutable until such time. The values in test were looked up first for a loaded environment and hardcoded for testing purposes only.